### PR TITLE
ci(test): migrate test cmd

### PR DESCRIPTION
**Pull request by ShaMan123**
This PR changes the `test` cmd to use `testem`.
This is done so we can stop using `fabricjs.com` for local browser testing and discontinue any associations between the repos.

I decided to keep going with `testem` for now.
We can always migrate to `karma` if we want to.

Later on we can remove the commands of the website... ~~or should I do that here and now?~~
~~Should we move all `testem` files to `tests`? I think it's better order. Their location is used only by `scripts/index`~~ **DONE**

## Files
- `workflows` - semantic cmd change
- `scripts/index` - refactor to support testem + put all local cache files in `.fabric` folder
- `test/GoldensServer.js` - extract from `scripts/index`, now runs when testem start visual tests
- `testem*` - adapted config/test page, moved under `test` and refactored from `json` to `js`
- `test/lib/*` - adapted methods
- `rollup.test.config.js` - need help here, can't get the tests to build properly

Logging has become silly because the program spawns tests in parallel (node, chrome, firefox) so it logs a mess when `--verbose` option is set. Well actually it is not that bad because ci runs only one context at a time so there will be no parallel logging there. Only locally, so I think it is fine. The `-c,--context` option can be used to run a single context. So I can't care less about this. And the test results are dumped to separate files under `.fabric` so the dev can view it there nice and orderly

## Changes

- tests log only errors (exposed `--verbose` flag)
image

- renamed `dev` cmd to `fabric`

### What about the rest of the interactive tests on fabricjs.com?

Started another effort migrating benchmarks, will PR in the near future

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -77,10 +75,12 @@
     "commander": "^9.1.0",
     "deep-object-diff": "^1.1.7",
     "eslint": "^8.21.0",
+    "fireworm": "^0.7.2",
```
Comment (ShaMan123) [inline package.json:78]: fixes `testem ci` error

**Review by ShaMan123 — COMMENTED** [review]
should I remove `old-travis-reference.yml`?

**Review by ShaMan123 — COMMENTED** [review]
Added support for `-r, -d` flags which make maintaining goldens super easy

**Comment by asturur** [comment]
We can continue to export tests until we have the new website. Then we can kill the test export

File: old-travis-reference.yml [hunk 2]
```diff
@@ -48,14 +48,14 @@ jobs:
...
+      env: LAUNCHER=chrome
+      script: npm run build:fast && npm run test -- -s unit -p 8080 -c $LAUNCHER
       addons:
         apt:
           packages: # avoid installing packages
     - stage: Unit Tests
-      env: LAUNCHER=Firefox
-      script: npm run build:fast && testem ci --port 8080 -f testem.json -l $LAUNCHER
+      env: LAUNCHER=firefox
+      script: npm run build:fast && npm run test -- -s unit -p 8080 -c $LAUNCHER
```
Comment (asturur) [inline old-travis-reference.yml:58]: i think this file is gone. we can trash it.

Comment (ShaMan123) [reply old-travis-reference.yml]: done

File: rollup.test.config.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import commonjs from '@rollup/plugin-commonjs';
+import ts from "rollup-plugin-ts";
+
+// https://rollupjs.org/guide/en/#configuration-files
+export default {
+  external: ['qunit', 'fabric'],
```
Comment (ShaMan123) [inline rollup.test.config.js:6]: not sure how to configure the tests
they build but don't start

File: rollup.test.config.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      format: "cjs",
+      exports: "named",
+      preserveModules: true, // Keep directory structure and files
+    },
+  ],
+  plugins: [
+		ts({
+			/* Plugin options */
+    }),
+    commonjs()
```
Comment (ShaMan123) [inline rollup.test.config.js:20]: needed for qunit

File: scripts/index.js [hunk 5]
```diff
@@ -136,8 +142,9 @@ function build(options = {}) {
         env: {
             ...process.env,
             MINIFY: Number(!options.fast),
-            BUILD_DESTINATION: options.destination,
-            MIN_BUILD_DESTINATION: minDest
+            BUILD_INPUT: options.input,
```
Comment (ShaMan123) [inline scripts/index.js:145]: good for devs who will want to do a custom build perhaps

File: scripts/index.js [hunk 6]
```diff
@@ -245,56 +252,114 @@ function exportToWebsite(options) {
...
+            fields[name] = value;
+        });
+        bb.on('close', () => {
+            resolve({ files, fields });
+        });
+        req.pipe(bb);
+    });
+}
+
+function startGoldensServer() {
```
Comment (ShaMan123) [inline scripts/index.js:278]: this is great!
it is proxied by testem and once a test fails it updates the golden file

File: scripts/index.js [hunk 7]
```diff
@@ -245,56 +252,114 @@ function exportToWebsite(options) {
...
+    const url = `http://localhost:${port}/`;
+    const start = (os.platform() == 'darwin' ? 'open' : os.platform() == 'win32' ? 'start' : 'xdg-open');
+    options.launch && cp.exec([start, url].join(' '));
+    // run
+    cp.spawn(args.join(' '), {
+        cwd: wd,
+        env: process.env,
+        shell: true,
+        stdio: 'pipe',
+        detached: !options.ci
```
Comment (ShaMan123) [inline scripts/index.js:357]: how do I close a detached process when the parent process dies?
It must be detached...

File: scripts/index.js [hunk 8]
```diff
@@ -434,12 +508,12 @@ program
     .description('build dist')
     .option('-f, --fast', 'skip minifying')
     .option('-w, --watch')
-    .option('-d, --dest [path]', 'specify the build destination path')
+    .option('-i, --input [...path]', 'specify the build input paths')
+    .option('-o, --output [path]', 'specify the build output path')
```
Comment (ShaMan123) [inline scripts/index.js:512]: renamed for consistency with rollup naming

File: scripts/index.js [hunk 9]
```diff
@@ -452,7 +526,11 @@ program
     .option('-d, --debug', 'debug visual tests by overriding refs (golden images) in case of visual changes', false)
     .option('-r, --recreate', 'recreate visual refs (golden images)', false)
     .option('-v, --verbose', 'log passing tests', false)
-    .option('-cc, --clear-cache', 'clear CLI test cache', false)
+    .option('-l, --launch', 'launch tests in the browser', false)
+    .option('--ci, --no-ci', 'runs testem with a `ci` flag', true)
```
Comment (ShaMan123) [inline scripts/index.js:530]: perhaps a better options would be `--dev` that is the opposite of `--ci`

Comment (ShaMan123) [reply scripts/index.js]: [f54ee83](https://github.com/fabricjs/fabric.js/pull/8138/commits/f54ee831cdbcbf681d66ff9a4abb73e8c749fe8a)

**Review by ShaMan123 — COMMENTED** [review]
@asturur I have left relevant comments in previous reviews
This is PR is **READY**
I would like to make tests es6 as well in an upcoming PR (don't know how to setup rollup config or whatever that is broken)
And bring `benchmarks` and `kitchensink` from the website - what do you think?
Then we can serve them with `testem` so debugging (including interactive testing) will be as simple as running `npm test`

**Comment by ShaMan123** [comment]
Maybe I should rename `.dumps` to `.fabric` 
much better...

**Comment by ShaMan123** [comment]
I think I should rename the `dev` cmd to something else (ci? tools? fabrici (fabric-ci)?) and make the it a cmd that builds and watches `src` and `test`

File: package.json [hunk 10]
```diff
@@ -46,41 +46,43 @@
   "scripts": {
     "changelog": "auto-changelog -o change-output.md --unreleased-only",
     "build": "node ./scripts build",
-    "build-rollup": "rollup -c",
     "build:fast": "npm run build -- --fast",
-    "dev": "node ./scripts",
+    "fabric": "node ./scripts",
```
Comment (ShaMan123) [inline package.json:50]: renamed!

File: scripts/GoldensServer.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
+const fs = require('fs-extra');
```
Comment (ShaMan123) [inline scripts/GoldensServer.js:1]: extracted from`scripts/index.js`

**Review by ShaMan123 — COMMENTED** [review]
refactored testem config files to `js` for ease of use and flexibility

I have changed default logging of tests to log only errors and added a `verbose` flag

**MORE THAN READY TO MERGE**

**Review by ShaMan123 — COMMENTED** [review]
ADDED dumping files that tests report to

**Comment by asturur** [comment]
I would like the .fabric folder with test_results to be only test_results.
Accessing the hidden folders with the . in front is a pain on a mac, as long as is git ignored we don't care for it to be hidden.

I'm also trying to move the coverage report under the same command but it doesn't seem to work

File: package.json [hunk 12]
```diff
@@ -53,6 +53,8 @@
     "export": "node ./scripts website export",
     "build-tests": "rollup -c ./rollup.test.config.js",
     "test": "node ./scripts test",
+    "test:unit-browser": "npm run test -- -s unit -p 8080 -l -c ",
+    "test:visual-browser": "npm run test -- -s visual -p 8080 -l -c ",
```
Comment (ShaMan123) [inline package.json:57]: `-c` flag is already taken

Comment (ShaMan123) [reply package.json:57]: make it `--coverage`

**Comment by ShaMan123** [comment]
I tried that and didn't get it to work

File: scripts/index.js [hunk 13]
```diff
@@ -429,30 +432,43 @@ program
...
     .option('-d, --debug', 'debug visual tests by overriding refs (golden images) in case of visual changes', false)
     .option('-r, --recreate', 'recreate visual refs (golden images)', false)
     .option('-v, --verbose', 'log passing tests', false)
-    .option('-cc, --clear-cache', 'clear CLI test cache', false)
+    .option('-l, --launch', 'launch tests in the browser', false)
+    .option('--dev', 'runs testem in `dev` mode, without a `ci` flag', false)
+    .addOption(new commander.Option('-c, --context <context...>', 'context to test in').choices(['chrome', 'firefox', 'node']).default(['chrome', 'node']))
+    .option('-p, --port')
+    .option('-o, --out <out>', 'path to report test results to')
+    .option('--clear-cache', 'clear CLI test cache', false)
```
Comment (ShaMan123) [inline scripts/index.js:471]: You should add the option `--coverage` here

Comment (ShaMan123) [reply scripts/index.js:471]: https://github.com/testem/testem#running-browser-code-after-tests-complete

Comment (ShaMan123) [reply scripts/index.js:471]: karma has it....
https://www.npmjs.com/package/karma-coverage

Comment (ShaMan123) [reply scripts/index.js:471]: https://github.com/testem/testem/blob/master/examples/coverage_istanbul/

Comment (ShaMan123) [reply scripts/index.js:471]: image
I managed to get it to work only on `node`
is that enough?

**Comment by asturur** [comment]
i wasn't able to move the coverage command, i always get a command that hangs

**Comment by asturur** [comment]
i wasn't able to move the coverage command, i always get a command that hangs

**Comment by asturur** [comment]
show you what i did
```diff
diff --git a/scripts/index.js b/scripts/index.js
index 546e5ec7..e9802a20 100644
--- a/scripts/index.js
+++ b/scripts/index.js
@@ -148,7 +148,7 @@ function startWebsite() {
     const args = ['run', 'start:dev'];
 
     //  WSL ubuntu
-    // https://github.com/microsoft/WSL/issues/216 
+    // https://github.com/microsoft/WSL/issues/216
     // os.platform() === 'win32' && args.push('--', '--force_polling', '--livereload');
     if (os.platform() === 'win32') {
         console.log(chalk.green('Consider using ubuntu on WSL to run jekyll with the following options:'));
@@ -272,7 +272,7 @@ async function test(suite, tests, options = {}) {
         env: {
             ...process.env,
             TEST_FILES: (tests || []).join(','),
-            NODE_CMD: ['qunit', 'test/node_test_setup.js', 'test/lib'].concat(tests || `test/${suite}`).join(' '),
+            NODE_CMD: [...(options.coverage ? ['nyc --silent']: []), 'qunit', 'test/node_test_setup.js', 'test/lib'].concat(tests || `test/${suite}`).join(' '),
             VERBOSE: Number(options.verbose),
             QUNIT_DEBUG_VISUAL_TESTS: Number(options.debug),
             QUNIT_RECREATE_VISUAL_REFS: Number(options.recreate),
@@ -469,6 +469,7 @@ program
     .option('-p, --port')
     .option('-o, --out <out>', 'path to report test results to')
     .option('--clear-cache', 'clear CLI test cache', false)
+    .option('-cov --coverage', 'Add coverage report with nyc', false)
     .action((options) => {
         if (options.clearCache) {
             fs.removeSync(CLI_CACHE);
@@ -538,4 +539,4 @@ program
         });
     });
 
-program.parse(process.argv);
\ No newline at end of file
+program.parse(process.argv);
```

i get a twst command that run super slow and make tests fail

**Comment by ShaMan123** [comment]
I did something pretty much the same.
Nothing lags on my machine

**Comment by ShaMan123** [comment]
```diff
- ...(options.coverage ? ['nyc --silent']: [])
+ ...options.coverage ? 'nyc --silent': ''
```

File: package.json (hunk 1 above)
Comment (asturur) [reply package.json:82]: but what is it?

File: package.json (hunk 12 above)
Comment (asturur) [reply package.json:57]: this is the -c flag to append chrome/firefox is not the coverage flag

File: scripts/index.js (hunk 13 above)
Comment (asturur) [reply scripts/index.js:471]: yes coverage needs only node, but results are weird in this screenshot.

Comment (ShaMan123) [reply scripts/index.js:471]: it's only for activeselection

File: package.json (hunk 12 above)
Comment (ShaMan123) [reply package.json:57]: so `-c chrome,firefox`

File: package.json (hunk 1 above)
Comment (ShaMan123) [reply package.json:82]: the error?
or the package?

Comment (ShaMan123) [reply package.json:82]: I forgot what was the error...

Comment (asturur) [reply package.json:82]: no this package. How does it fix n error?

**Review by asturur — APPROVED** [review]

**Pull request merged**